sql: added Hidden to index descriptor and SHOW INDEX

wenyihu6 opened the pull request:
This PR takes the first step for the invisible index feature. It added a boolean
field `Hidden` to the struct `IndexDescriptor`. Since primary indexes cannot be
hidden, this pr also added a check in `pkg/sql/catalog/tabledesc/validate.go`
for that. In addition, this PR also added a new column `visible` to
`crdb_internal.table_indexes` and also to the output of following SQL statements:
```
SHOW INDEX FROM (table_name)
SHOW INDEXES FROM(table_name)
SHOW KEYS FROM (table_name)
SHOW INDEX FROM DATABASE(database_name)
SHOW INDEXES FROM DATABASE (database_name)
SHOW KEYS FROM DATABASE (database_name)
```
Since the invisible index feature has not been introduced yet, all indexes
created should be visible. It is expected for all test cases to output `true`
for all `visible` columns.

See also: next PR for invisible index feature: https://github.com/cockroachdb/cockroach/pull/83471

Assists: https://github.com/cockroachdb/cockroach/issues/72576
Note that this issue has not been resolved yet, and this pr only takes the first step.

Release note (sql change): A new column `visible` has been added to the table
`crdb_internal.table_indexes` and to the output of SQL statements related to
`SHOW INDEX`, `SHOW INDEXES`, and `SHOW KEYS`. The `visible` column indicates
whether the index is visible to the optimizer. An invisible index is an index
that is up-to-date but is ignored by the optimizer unless explicitly specified
with [index hinting](https://www.cockroachlabs.com/docs/v22.1/table-expressions#force-index-selection).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83388)
<!-- Reviewable:end -->

postamar commented on pkg/sql/alter_primary_key.go:
```diff
@@ -183,6 +183,7 @@ func (p *planner) AlterPrimaryKey(
 		Version:        descpb.StrictIndexColumnIDGuaranteesVersion,
 		ConstraintID:   tableDesc.GetNextConstraintID(),
 		CreatedAtNanos: p.EvalContext().GetTxnTimestamp(time.Microsecond).UnixNano(),
+		Invisible:      false,
```
Naming consistency: consider renaming to `Hidden`.

postamar commented on pkg/sql/alter_table.go:
```diff
@@ -280,6 +280,7 @@ func (n *alterTableNode) startExec(params runParams) error {
 				idx := descpb.IndexDescriptor{
 					Name:             string(d.Name),
 					Unique:           true,
+					Invisible:        false,
```
Naming consistency: consider renaming the field in the index descriptor proto to `hidden`.

postamar commented on pkg/sql/catalog/catformat/index.go:
```diff
@@ -174,6 +174,10 @@ func indexForDisplay(
 	}
 
+	if index.Invisible {
+		f.WriteString(" INVISIBLE")
```
Naming consistency: consider replacing the `INVISIBLE` SQL keyword with `NOT VISIBLE`.

postamar commented on pkg/sql/catalog/descpb/index.go:
```diff
@@ -28,6 +28,11 @@ func (desc *IndexDescriptor) IsPartial() bool {
+// IsInvisible returns whether the index is invisible or not.
+func (desc *IndexDescriptor) IsInvisible() bool {
+	return desc.Invisible
+}
```
Is this method necessary? This isn't super obvious but we try to keep the generated protobuf structs lightweight and instead move as much business logic as possible to the `catalog.Index` interface and its implementation.

postamar commented on pkg/ccl/logictestccl/testdata/logic_test/crdb_internal_tenant:
```diff
@@ -280,10 +280,10 @@ SELECT * FROM crdb_internal.table_columns WHERE descriptor_name = ''
 SELECT * FROM crdb_internal.table_indexes WHERE descriptor_name = ''
 ----
-descriptor_id  descriptor_name  index_id  index_name  index_type  is_unique  is_inverted  is_sharded shard_bucket_count created_at
+descriptor_id  descriptor_name  index_id  index_name  index_type  is_unique  is_inverted  is_sharded is_invisible shard_bucket_count created_at
```
I posted a few other comments in the same vein, but they aren't exhaustive.

postamar commented on pkg/sql/catalog/tabledesc/validate_test.go:
```diff
@@ -652,6 +653,36 @@ func TestValidateTableDesc(t *testing.T) {
+				NextColumnID: 2,
+				NextFamilyID: 1,
+				NextIndexID:  2,
+			}},
```
Thanks for adding this test 👍 

Descriptor validation helps us prevent bugs which otherwise are time-consuming to fix, by catching inconsistencies before they are committed to the cluster.

postamar commented on pkg/sql/delegate/show_database_indexes.go:
```diff
@@ -16,34 +16,48 @@ import (
 	storing::BOOL,
-	implicit::BOOL`
+	implicit::BOOL,
+  idx.is_invisible::BOOL`
```
Consider `idx.is_invisible::BOOL AS hidden` instead to keep the column names consistent, not just with the `invisible -> hidden` rename I'm advocating, but also for consistency within this table. The other boolean columns don't have an `is_` prefix here.

postamar commented on pkg/sql/delegate/show_table.go:
```diff
@@ -111,7 +113,8 @@ SELECT
     storing::BOOL,
-    implicit::BOOL`
+    implicit::BOOL,
+    idx.is_invisible::BOOL`
```
Same as previous comment: `idx.is_invisible::BOOL AS hidden`

postamar commented on pkg/sql/opt/cat/index.go:
```diff
@@ -56,6 +56,9 @@ type Index interface {
 	IsInverted() bool
 
+	// IsInvisible returns true if this is an invisible index.
+	IsInvisible() bool
```
Naming consistency: consider renaming to `IsHidden()`.

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -214,6 +214,10 @@ message Index {
+
+  // is_invisible specifies whether this index is invisible. Note that primary
+  // index cannot be invisible.
+  bool is_invisible = 23;
```
In addition to suggesting to renaming this for consistency, could you please also add a TODO comment stating that this field will need to be populated by `scdecomp`?

postamar reviewed (commented):
Reviewing for schema, most of my comments are about the names of things, other than that this looks like a reasonable change, done very thoroughly. Thank you for that.

michae2 reviewed (commented):
This is looking really good! I only have a few comments.

> Reviewed 19 of 57 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-:-N6u2RTSCD2nbzF_ZHdj:b-v1jc3n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @postamar, and @wenyihu6)

___
*[`pkg/sql/alter_primary_key.go` line 196 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6tvOsIDYKfNBO-m8BH:-N6tvOsIDYKfNBO-m8BI:b-nzdcdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c3aed09fcdb745116aa3b8efea1a1588d8b1d/pkg/sql/alter_primary_key.go#L196)):*
> ```Go
> 		ConstraintID:   tableDesc.GetNextConstraintID(),
> 		CreatedAtNanos: p.EvalContext().GetTxnTimestamp(time.Microsecond).UnixNano(),
> 		Hidden:         false,
> ```

nit: We don't actually need to specify `Hidden: false` in all of these `IndexDescriptor` struct literals. Go has a nice feature where fields omitted from struct literals will get their zero value, and in the case of `bool` that is `false`. So we don't need to touch all these files to set `Hidden: false`; it will happen automatically.
___
*[`pkg/sql/catalog/descpb/structured.proto` line 305 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6ttEcmC18kAT462d8S:-N6ttEcmC18kAT462d8T:bhg2lno) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c3aed09fcdb745116aa3b8efea1a1588d8b1d/pkg/sql/catalog/descpb/structured.proto#L305)):*
> ```protobuf
> //   key_column_directions:  ASC, ASC
> //   key_column_ids:         1, 2   // k1, k2
> //   hidden: false
> ```

nit: Indentation is off here. And to be honest, I don't think we need to include `hidden: false` in these examples.
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 1295 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6u-B495FbjPDTKmwv0:-N6u-B495FbjPDTKmwv1:b-319c39) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c3aed09fcdb745116aa3b8efea1a1588d8b1d/pkg/sql/catalog/tabledesc/validate.go#L1295)):*
> ```Go
> 			}
> 			if idx.IsHidden() {
> 				return errors.Newf("a primary index %q cannot be hidden.",
> ```

nit: You can drop "a " from the error message.
___
*[`pkg/sql/delegate/show_database_indexes.go` line 40 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N5b46mPwqiCkPSPygfc-r3-40:-N6tuKgjBP-BhCNaPWQ7:by79hqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b659287390d5e01b9db1ed045fbe5bf1a55f3169/pkg/sql/delegate/show_database_indexes.go#L40)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider `idx.is_invisible::BOOL AS hidden` instead to keep the column names consistent, not just with the `invisible -> hidden` rename I'm advocating, but also for consistency within this table. The other boolean columns don't have an `is_` prefix here.
</blockquote></details>

Since the `SHOW INDEXES` statement comes from MySQL, and this output is already different from `crdb_internal.table_indexes`, my vote is that we might as well match them here and below by using `visible` as the column name (and inverting the meaning).
___
*[`pkg/sql/opt/exec/execbuilder/testdata/explain` line 633 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6u17xk9ivKrVZp5GeV:-N6u17xk9ivKrVZp5GeW:b-wxy4kj) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c3aed09fcdb745116aa3b8efea1a1588d8b1d/pkg/sql/opt/exec/execbuilder/testdata/explain#L633)):*
> ```
>             │
>             └── • virtual table
>                   table: statistics@primary
> ```

Hmm, this shouldn't be changing. :thinking: Maybe this slipped in from a rebase?


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
Few more comments. This is almost there! :smile_cat:

> Reviewed 1 of 57 files at r4, 2 of 34 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-:-N6zBGxu-DfpFA_sNgdG:b-m71eqc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, @postamar, and @wenyihu6)

___
*[`-- commits` line 22 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6zB71L9I53x58q39ir):*
nit: let's remove the todo
___
*[`-- commits` line 24 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6zA4pZAjoqXySOblYL):*
nit: When you put "Fixes" in the PR, GitHub will [automatically close](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) the issue. For this first PR, instead of "Fixes" I would use "Informs" or "Assists" or "Helps" or some other verb, since we don't want to close the issue yet.
___
*[`-- commits` line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6zAhXZ2ePuh5z1D5Mw):*
nit: You can leave off the last bit about "Since the invisible index feature has not been introduced yet...". By the time people read the release notes, the feature will be totally done! :slightly_smiling_face:
___
*[`pkg/sql/crdb_internal.go` line 2783 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6z894i3a78lm-8gQ2n:-N6z894i3a78lm-8gQ2o:bnc0uol) ([raw file](https://github.com/cockroachdb/cockroach/blob/e557cb73d3bf3797a7148d194147d71fb6553706/pkg/sql/crdb_internal.go#L2783)):*
> ```Go
>   is_inverted         BOOL NOT NULL,
>   is_sharded          BOOL NOT NULL,
>   visible             BOOL NOT NULL,
> ```

Oh, I didn't mean for you to change this one, sorry! Let's keep the `is_hidden` column name here in `crdb_internal.table_indexes` (and in all other `crdb_internal` tables) and use `NOT idx.is_hidden::BOOL AS visible` in `delegateShowDatabaseIndexes` and `delegateShowIndexes`.
___
*[`pkg/sql/delegate/show_database_indexes.go` line 61 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6z6KNw2kV7cUId-C83:-N6z6KNw2kV7cUId-C84:b4qr7iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e557cb73d3bf3797a7148d194147d71fb6553706/pkg/sql/delegate/show_database_indexes.go#L61)):*
> ```Go
>       idx.index_name = info.index_name AND 
> 			idx.descriptor_name = info.table_name`
> ```

Let's add an order by to this query as well:

```
getAllIndexesQuery += `
ORDER BY
    1, 2, 4`
```

(And then you won't need all the `ORDER BY` in the logic tests, although they're also fine to have, aren't hurting anything.)
___
*[`pkg/sql/delegate/show_table.go` line 146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6z743qCyXSS1Zgardv:-N6z743qCyXSS1Zgardw:b-nes6px) ([raw file](https://github.com/cockroachdb/cockroach/blob/e557cb73d3bf3797a7148d194147d71fb6553706/pkg/sql/delegate/show_table.go#L146)):*
> ```Go
>     AND table_name=%[2]s
> ORDER BY
>     1, 2, 3, 4, 5, 6, 7, 8, 9;`
> ```

I know you didn't write it, but while we're here, this order by can be simplified to:

```
ORDER BY
    1, 2, 4`
```
___
*[`pkg/sql/opt/exec/execbuilder/testdata/explain` line 633 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6u17xk9ivKrVZp5GeV:-N6z9i4-3NzLaEc_Hvwq:b-mwvjgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c3aed09fcdb745116aa3b8efea1a1588d8b1d/pkg/sql/opt/exec/execbuilder/testdata/explain#L633)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Hmm, this shouldn't be changing. :thinking: Maybe this slipped in from a rebase?
</blockquote></details>

Oh, I'm wrong, sorry. :flushed: Ignore me. :slightly_smiling_face:


<!-- Sent from Reviewable.io -->

michae2 commented:
When you've addressed comments and you're feeling ready, you'll also need to rebase this on the tip of master to fix the conflicts. (For multi-PR rebases, I usually do `git rebase master` for the first PR's branch, and then after that `git rebase -i <first PR branch>` for the second PR's branch, `git rebase -i <second PR branch>` for the third PR's branch, etc.)

wenyihu6 commented:
I see. Thanks for reviewing my PR and the explanation here!





<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`-- commits` line 22 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6zB71L9I53x58q39ir):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: let's remove the todo
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`-- commits` line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6zAhXZ2ePuh5z1D5Mw):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: You can leave off the last bit about "Since the invisible index feature has not been introduced yet...". By the time people read the release notes, the feature will be totally done! :slightly_smiling_face:
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/alter_table.go` line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N5VydtlnYu--fVsIjqq-r3-283:-N70WaYhBJdF5xePQFBn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b659287390d5e01b9db1ed045fbe5bf1a55f3169/pkg/sql/alter_table.go#L283)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Naming consistency: consider renaming the field in the index descriptor proto to `hidden`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/crdb_internal.go` line 2783 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6z894i3a78lm-8gQ2n:-N70Weno1yX6CALHt7zA:bh7pzsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e557cb73d3bf3797a7148d194147d71fb6553706/pkg/sql/crdb_internal.go#L2783)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Oh, I didn't mean for you to change this one, sorry! Let's keep the `is_hidden` column name here in `crdb_internal.table_indexes` (and in all other `crdb_internal` tables) and use `NOT idx.is_hidden::BOOL AS visible` in `delegateShowDatabaseIndexes` and `delegateShowIndexes`.
</blockquote></details>

Ahh I see why we want `is_hidden`in `crdb_internal` now. Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/catalog/catformat/index.go` line 178 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N5VydtlnYu--fVsIjqr-r3-178:-N70X3sZCAlg-YIEL3v_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b659287390d5e01b9db1ed045fbe5bf1a55f3169/pkg/sql/catalog/catformat/index.go#L178)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Naming consistency: consider replacing the `INVISIBLE` SQL keyword with `NOT VISIBLE`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/catalog/tabledesc/validate.go` line 1295 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6u-B495FbjPDTKmwv0:-N70X97j8_AQEaYBvaj5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c3aed09fcdb745116aa3b8efea1a1588d8b1d/pkg/sql/catalog/tabledesc/validate.go#L1295)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: You can drop "a " from the error message.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/delegate/show_database_indexes.go` line 61 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6z6KNw2kV7cUId-C83:-N70XMM84Fb7TDR5Vwu0:bxggei0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e557cb73d3bf3797a7148d194147d71fb6553706/pkg/sql/delegate/show_database_indexes.go#L61)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Let's add an order by to this query as well:

```
getAllIndexesQuery += `
ORDER BY
    1, 2, 4`
```

(And then you won't need all the `ORDER BY` in the logic tests, although they're also fine to have, aren't hurting anything.)
</blockquote></details>

Ahh that's a good idea. Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/catalog/descpb/index.go` line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N5VydtlnYu--fVsIjqs-r3-34:-N70Xe2k-4AWlgmWBdfy:bbc5e0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b659287390d5e01b9db1ed045fbe5bf1a55f3169/pkg/sql/catalog/descpb/index.go#L34)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this method necessary? This isn't super obvious but we try to keep the generated protobuf structs lightweight and instead move as much business logic as possible to the `catalog.Index` interface and its implementation.
</blockquote></details>

Hmm I see. It's not necessary here. This function is not called from anywhere else. I just removed it. 



_Code quote:_
```Go
func (desc *IndexDescriptor) IsHidden() bool {
	return desc.Hidden
}
```


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/delegate/show_table.go` line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N5b46mPwqiCkPSPygfd-r3-117:-N70Xux3EpkTgDIrBWpu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b659287390d5e01b9db1ed045fbe5bf1a55f3169/pkg/sql/delegate/show_table.go#L117)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Same as previous comment: `idx.is_invisible::BOOL AS hidden`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/delegate/show_table.go` line 146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6z743qCyXSS1Zgardv:-N70YIDwAX6L7e0g_Eq1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e557cb73d3bf3797a7148d194147d71fb6553706/pkg/sql/delegate/show_table.go#L146)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I know you didn't write it, but while we're here, this order by can be simplified to:

```
ORDER BY
    1, 2, 4`
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/cat/index.go` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N5VydtmOkKd8fKo1-q8-r3-60:-N70YLSUB94-y02iXtLO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b659287390d5e01b9db1ed045fbe5bf1a55f3169/pkg/sql/opt/cat/index.go#L60)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Naming consistency: consider renaming to `IsHidden()`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/ccl/logictestccl/testdata/logic_test/crdb_internal_tenant` line 286 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N5b46mPwqiCkPSPygfa-r3-286:-N70YQ99CrCpMC5bAjuI:baann2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b659287390d5e01b9db1ed045fbe5bf1a55f3169/pkg/ccl/logictestccl/testdata/logic_test/crdb_internal_tenant#L286)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I posted a few other comments in the same vein, but they aren't exhaustive.
</blockquote></details>

Yeah. I just changed the column to `is_hidden` in crdb_internal.table_indexes, and `visible` for all `SHOW INDEX` related SQL statements. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/explain` line 633 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6u17xk9ivKrVZp5GeV:-N70Ycr476qK7pkYUvhj:b-9essue) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c3aed09fcdb745116aa3b8efea1a1588d8b1d/pkg/sql/opt/exec/execbuilder/testdata/explain#L633)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Oh, I'm wrong, sorry. :flushed: Ignore me. :slightly_smiling_face:
</blockquote></details>

Was the changes from the order_by?


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/alter_primary_key.go` line 196 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N6tvOsIDYKfNBO-m8BH:-N70YwHA93rZ7Xp4ZqdW:b-kqra99) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c3aed09fcdb745116aa3b8efea1a1588d8b1d/pkg/sql/alter_primary_key.go#L196)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: We don't actually need to specify `Hidden: false` in all of these `IndexDescriptor` struct literals. Go has a nice feature where fields omitted from struct literals will get their zero value, and in the case of `bool` that is `false`. So we don't need to touch all these files to set `Hidden: false`; it will happen automatically.
</blockquote></details>

Thanks for the explanation! I removed some `Hidden: false`, but kept some `Hidden: false` to be explicit at some places to remind myself that I later need to populate the data correctly after the syntax is introduced. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit` line 517 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N70fDmJ-tWegsVbLc48:-N70fDmJ-tWegsVbLc49:bqcrev9) ([raw file](https://github.com/cockroachdb/cockroach/blob/86e007dcb5cbde3501f138c1d768519db3487857/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit#L517)):*
> ```
> query TTB
> SELECT index_name, column_name, implicit FROM [SHOW INDEXES FROM t]
> ORDER BY index_name, seq_in_index
> ```

I removed `ORDER BY index_name, seq_in_index` from some test cases because it's now using `ORDER BY table_name, index_name, seq_in_index` in [SHOW INDEXES](https://github.com/cockroachdb/cockroach/blob/7de496578d9f9cd2090953676eecace256110f2b/pkg/sql/delegate/show_table.go#L146). Let me know if you want me to add them back.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit` line 517 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83388#-N70fDmJ-tWegsVbLc48:-N70hvtP3Kyxs6N2ELUC:b-6ytf3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/86e007dcb5cbde3501f138c1d768519db3487857/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit#L517)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I removed `ORDER BY index_name, seq_in_index` from some test cases because it's now using `ORDER BY table_name, index_name, seq_in_index` in [SHOW INDEXES](https://github.com/cockroachdb/cockroach/blob/7de496578d9f9cd2090953676eecace256110f2b/pkg/sql/delegate/show_table.go#L146). Let me know if you want me to add them back.
</blockquote></details>

Hmm I'm not sure why the changes I made is not showing here in Reviewable. But they are under `FilesChanged` in my [PR](https://github.com/cockroachdb/cockroach/pull/83388/files).


<!-- Sent from Reviewable.io -->

mgartner commented:
@postamar apologies for re-opening this discussion, but I'm worried that using the word "hidden" for parts of the feature when all of the user-facing interactions with the feature use the term "visible" make this unnecessarily confusing. Do you feel that naming consistent with existing "hidden" features is more important than consistency with the "not visible" feature being added?

wenyihu6 commented:
@postamar @mgartner @michae2 @knz @vy-ton @otan Tagging people who have commented on the syntax discussion so far. Feel free to ignore me if this is not in your zone : ) 

It’s been a long time since we last talked about the syntax and naming conventions for this feature. Sorry about this! This paragraph summarizes our discussion about this topic so far.

Invisible index feature is introducing three new user facing syntaxes. When users are creating a new invisible index with `CREATE TABLE` or with `CREATE INDEX`, they will need to specify whether the index is invisible. The two options that we have discussed are `NOT VISIBLE`, `INVISIBLE`. [MySQL](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html) and [Oracle](https://oracle-base.com/articles/11g/invisible-indexes-11gr1) both support `INVISIBLE` instead. Invisible column feature is using `NOT VISIBLE`. But we have decided that being consistent with the invisible column feature is important.

The second user-facing syntax is related to SQL statements like`SHOW INDEX`. The three options are `is_hidden`, `visible`, or `is_visible`. [Oracle](https://docs.oracle.com/cd/E17952_01/mysql-8.0-en/invisible-indexes.html) is using `is_visible`. [MySQL](https://dev.mysql.com/doc/refman/8.0/en/show-index.html) is using `visible`. Invisible column feature is using `is_hidden`. And now we want to decide if it is more important to stay consistent with the invisible column feature and use `is_hidden` or to stay consistent with the first user-facing syntax and use `is_visible`. I slightly prefer to use `is_visible` over `visible` to be consistent with other columns in SHOW INDEX — `is_unique`, `is_inverted`.

The third user-facing syntax is with `crdb_internal.table_indexes`. Invisible column feature is using`is_hidden` for` table_indexes`. Oracle, MySQL are both consistent with what they have for `SHOW INDEX` which are `is_visible` and `visible`.

We are also introducing another field in the index descriptor. The options are `hidden` or `invisible`. The invisible column feature is using `hidden` in column descriptor.  If we decide to choose `visible` for all invisible index syntax mentioned above, we can use `invisible` here to be more consistent. My slight preference is to use either hidden or invisible (not visible) for the index descriptor since the default behavior for a new index is visible, and the default boolean value is false.

Another thing to note is that MySQL and Oracle both introduce a new column in `INFORMATION_SCHEMA.STATISTICS`. We are now introducing a new column in  `crdb_internal.table_indexes`, but we can add a new column to `INFORMATION_SCHEMA.STATISTICS` as well.





<!-- Sent from Reviewable.io -->

michae2 commented:
Naming is hard! 🙂

> Invisible index feature is introducing three new user facing syntaxes. When users are creating a new invisible index with `CREATE TABLE` or with `CREATE INDEX`, they will need to specify whether the index is invisible. The two options that we have discussed are `NOT VISIBLE`, `INVISIBLE`. [MySQL](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html) and [Oracle](https://oracle-base.com/articles/11g/invisible-indexes-11gr1) both support `INVISIBLE` instead. Invisible column feature is using `NOT VISIBLE`. But we have decided that being consistent with the invisible column feature is important.

Yes, I agree with the current usage of `NOT VISIBLE` in DDL statements. It matches what we do for columns and it avoids the problems with making `INVISIBLE` a keyword in the grammar. It's too bad it doesn't match MySQL and Oracle, but I've accepted that. 😛

> The second user-facing syntax is related to SQL statements like`SHOW INDEX`. The three options are `is_hidden`, `visible`, or `is_visible`. [Oracle](https://docs.oracle.com/cd/E17952_01/mysql-8.0-en/invisible-indexes.html) is using `is_visible`. [MySQL](https://dev.mysql.com/doc/refman/8.0/en/show-index.html) is using `visible`. Invisible column feature is using `is_hidden`. And now we want to decide if it is more important to stay consistent with the invisible column feature and use `is_hidden` or to stay consistent with the first user-facing syntax and use `is_visible`. I slightly prefer to use `is_visible` over `visible` to be consistent with other columns in SHOW INDEX — `is_unique`, `is_inverted`.

The `SHOW INDEX` statements are not standard, and come from [MySQL](https://dev.mysql.com/doc/refman/8.0/en/show-index.html), so I propose we match MySQL and use `visible` in the output of these statements. (Oracle doesn't actually support `SHOW INDEX`, that link is also for MySQL.)

> The third user-facing syntax is with `crdb_internal.table_indexes`. Invisible column feature is using`is_hidden` for` table_indexes`. Oracle, MySQL are both consistent with what they have for `SHOW INDEX` which are `is_visible` and `visible`.

I don't feel strongly about this one. I'm fine with `is_hidden` to match what we do for columns. Or anything else.

> We are also introducing another field in the index descriptor (just for internal use). The options are `hidden`, `invisible`. The invisible column feature is using `hidden` in column descriptor. If we decide to choose `visible` for all invisible index syntax mentioned above, we can use `invisible` here to be more consistent. My slight preference is to use either hidden or invisible (not visible) for the index descriptor since the default behavior for a new index is visible, and the default boolean value is false.

Again, I don't feel strongly about this one. I'm fine with `hidden` or `invisible`. Anything is fine with me.

> Another thing to note is that MySQL and Oracle both introduce a new column in `INFORMATION_SCHEMA.STATISTICS`. We are now introducing a new column in `crdb_internal.table_indexes`, but we can add a new column to `INFORMATION_SCHEMA.STATISTICS` as well.

[MySQL](https://dev.mysql.com/doc/refman/8.0/en/information-schema-statistics-table.html) uses `is_visible` in the `STATISTICS` table, as an extension to the standard. (Oracle doesn't actually support `INFORMATION_SCHEMA`.) Postgres doesn't seem to have `information_schema.statistics`. Again, I propose we match MySQL.

wenyihu6 commented:
> Oracle



> Naming is hard! 🙂
> 
> > Invisible index feature is introducing three new user facing syntaxes. When users are creating a new invisible index with `CREATE TABLE` or with `CREATE INDEX`, they will need to specify whether the index is invisible. The two options that we have discussed are `NOT VISIBLE`, `INVISIBLE`. [MySQL](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html) and [Oracle](https://oracle-base.com/articles/11g/invisible-indexes-11gr1) both support `INVISIBLE` instead. Invisible column feature is using `NOT VISIBLE`. But we have decided that being consistent with the invisible column feature is important.
> 
> Yes, I agree with the current usage of `NOT VISIBLE` in DDL statements. It matches what we do for columns and it avoids the problems with making `INVISIBLE` a keyword in the grammar. It's too bad it doesn't match MySQL and Oracle, but I've accepted that. 😛
> 
> > The second user-facing syntax is related to SQL statements like`SHOW INDEX`. The three options are `is_hidden`, `visible`, or `is_visible`. [Oracle](https://docs.oracle.com/cd/E17952_01/mysql-8.0-en/invisible-indexes.html) is using `is_visible`. [MySQL](https://dev.mysql.com/doc/refman/8.0/en/show-index.html) is using `visible`. Invisible column feature is using `is_hidden`. And now we want to decide if it is more important to stay consistent with the invisible column feature and use `is_hidden` or to stay consistent with the first user-facing syntax and use `is_visible`. I slightly prefer to use `is_visible` over `visible` to be consistent with other columns in SHOW INDEX — `is_unique`, `is_inverted`.
> 
> The `SHOW INDEX` statements are not standard, and come from [MySQL](https://dev.mysql.com/doc/refman/8.0/en/show-index.html), so I propose we match MySQL and use `visible` in the output of these statements. (Oracle doesn't actually support `SHOW INDEX`, that link is also for MySQL.)
> 
> > The third user-facing syntax is with `crdb_internal.table_indexes`. Invisible column feature is using`is_hidden` for` table_indexes`. Oracle, MySQL are both consistent with what they have for `SHOW INDEX` which are `is_visible` and `visible`.
> 
> I don't feel strongly about this one. I'm fine with `is_hidden` to match what we do for columns. Or anything else.
> 
> > We are also introducing another field in the index descriptor (just for internal use). The options are `hidden`, `invisible`. The invisible column feature is using `hidden` in column descriptor. If we decide to choose `visible` for all invisible index syntax mentioned above, we can use `invisible` here to be more consistent. My slight preference is to use either hidden or invisible (not visible) for the index descriptor since the default behavior for a new index is visible, and the default boolean value is false.
> 
> Again, I don't feel strongly about this one. I'm fine with `hidden` or `invisible`. Anything is fine with me.
> 
> > Another thing to note is that MySQL and Oracle both introduce a new column in `INFORMATION_SCHEMA.STATISTICS`. We are now introducing a new column in `crdb_internal.table_indexes`, but we can add a new column to `INFORMATION_SCHEMA.STATISTICS` as well.
> 
> [MySQL](https://dev.mysql.com/doc/refman/8.0/en/information-schema-statistics-table.html) uses `is_visible` in the `STATISTICS` table, as an extension to the standard. (Oracle doesn't actually support `INFORMATION_SCHEMA`.) Postgres doesn't seem to have `information_schema.statistics`. Again, I propose we match MySQL.

I see now. Sorry about the confusion. The source I got for `Oracle` is actually for `MySQL`😬. I will correct that paragraph now.

vy-ton commented:
> The two options that we have discussed are NOT VISIBLE, INVISIBLE. [MySQL](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html) and [Oracle](https://oracle-base.com/articles/11g/invisible-indexes-11gr1) both support INVISIBLE instead. Invisible column feature is using NOT VISIBLE. But we have decided that being consistent with the invisible column feature is more important.

Can we support `INVISIBLE` as an alias? We've done this for other features where other dbs offer different syntax. The value being that an existing migration wouldn't need to change their SQL.

> The second user-facing syntax is related to SQL statements like SHOW INDEX. 
I prefer `visible`. I'm confused by `I slightly prefer to use is_visible over visible to be consistent with other columns in SHOW INDEX — is_unique, is_inverted.` since I see `non-unique` on SHOW INDEX.

We should help users understand the difference between hidden columns vs indexes, which is why I prefer the different naming for SHOW. For columns, being hidden is a presentation decision. For indexes, being hidden impacts query behavior.

* MySQL uses [INFORMATION_SCHEMA.STATISTICS](https://dev.mysql.com/doc/refman/8.0/en/information-schema-statistics-table.html) and has is_visible. - I prefer this. After Postgres syntax/semantics, I would favor mirroring MySQL.
* Yes, we should add the matching `information_schema` columns

wenyihu6 commented:
> 

Thanks for looking into this! We can choose `visible` for `SHOW INDEX` and `is_visible` for both `crdb_internal.table_indexes` and `information_schema`. Sorry, I mixed up `SHOW INDEX` and `crdb_internal.table_indexes`. For the alias part, I will try looking into it.

knz commented:
> Can we support INVISIBLE as an alias? We've done this for other features where other dbs offer different syntax. The value being that an existing migration wouldn't need to change their SQL.

I don't think that's a good idea. There are two different arguments, one that's user-facing and one that's technical.

1. For one, the reason "An existing migration wouldn't need to change their SQL" is not very good. If the user is coming from MySQL or Oracle, they have to rewrite their SQL anyway. So having different keywords is part of the course.

2.  the technical reason is that we should remain consistent on UX. We already support NOT VISIBLE for another feature (hidden columns). If we introduce INVISIBLE as alias for NOT VISIBLE in the new feature, we'd need to do it for hidden columns too for consistency. **However, we cannot do this for hidden columns** (for a technical reason that has to do with the structure of our grammar). Because we cannot do it for hidden columns, we should not do it here either.

mgartner commented:
I agree with @knz that adding `INVISIBLE` would only add confusion when users try to create non-visible columns, much like [the confusion with `FOR` and `FROM`](https://github.com/cockroachdb/cockroach/issues/57638) that we are already guilty of creating.

I agree with the others on the columns for introspection tables. Avoiding "invisible" and favoring column names with the word "visible" has the benefit of matching the syntax. We may consider referring to the feature publicly (i.e. in docs) as "index visibility" rather than "invisible indexes" to avoid inconsistency with the syntax.

vy-ton commented:
For migrations, it would be better UX if users could focus SQL rewrites on unsupported syntax or different db requirements, e.g. primary key design. Regardless, I understand the rationale for no alias. @otan Can you open an issue for creating a schema migration tool rule to convert syntax.

wenyihu6 commented:
Thanks everyone for the comments and sorry for the email spam!

Conclusion on invisible index syntax discussion: Michael, Vy, and Marcus's opinions on the second and third user-facing syntax are the same. SQL statements related to `SHOW INDEX` will use `visible`. `crdb_internal.table_indexes` will use `is_visible`. This is also not disagreed by anyone from our previous discussion; I will use them as the final decision if no one has different opinions. Please let me know if you do : )

For the first user-facing syntax, Vy wants to support `INVISIBLE` as an alias. I will leave this for discussion later. If we do want to support `INVISIBLE` as an alias later, I will make another PR in the end after I complete this invisible index feature. 

I will also make another PR to add a new column in `information_schema.statistics` with `is_visible` after this PR is merged.





<!-- Sent from Reviewable.io -->

vy-ton commented:
> For the first user-facing syntax, Vy wants to support INVISIBLE as an alias. I will leave this for discussion later. If we do want to support INVISIBLE as an alias later, I will make another PR in the end after I complete this invisible index feature.

Sorry I should have clarified that I'm ok with no support for `INVISIBLE` as alias

postamar commented:
> @postamar apologies for re-opening this discussion, but I'm worried that using the word "hidden" for parts of the feature when all of the user-facing interactions with the feature use the term "visible" make this unnecessarily confusing. Do you feel that naming consistent with existing "hidden" features is more important than consistency with the "not visible" feature being added?

I don't feel too strongly about that. I'm biased towards maintaining towards existing patterns and addressing the naming inconsistencies already present with columns falls way out of scope of this PR. Furthermore, that's neither urgent nor difficult, so I doesn't worry me.

wenyihu6 committed:
sql: added NotVisible to index descriptor

This PR added a new field `NotVisible` to the struct `IndexDescriptor`. Since
primary indexes cannot be not visible, it added another test in
`pkg/sql/catalog/tabledesc/validate.go`. Since the invisible index feature has
not been introduced yet, all indexes created should be visible.

See also: https://github.com/cockroachdb/cockroach/pull/83471

Assists: https://github.com/cockroachdb/cockroach/issues/72576

Release note: none

wenyihu6 commented:
Opened a new PR for a cleaner thread without syntax discussion (New pr: https://github.com/cockroachdb/cockroach/pull/84763, https://github.com/cockroachdb/cockroach/pull/84776).
